File: python/app/adapter/driven/user_persistence_adapter.py
import os
import sqlite3

from app.adapter.driven.query_exception import QueryException
from app.adapter.driven.repository_exception import RepositoryException
from app.application.user_dto import UserDto
from app.application.user_query_criteria import UserQueryCriteria
from app.application.user_query_handler import UserQueryHandler
from app.application.user_query_sort import UserQuerySort
from app.domain.user import User
from app.domain.user_repository import UserRepository
from app.domain.user_unique_checker import UserUniqueChecker
from app.utility import value_utility


class UserPersistenceAdapter(UserRepository, UserUniqueChecker, UserQueryHandler):

    def __init__(self):
        self.conn = sqlite3.connect(os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), 'db.db'))
        self.conn.row_factory = sqlite3.Row
        self.cursor = self.conn.cursor()
        self.cursor.execute(
            """
            create table if not exists t_user (u_id text primary key, u_username text, u_password text,
            u_nickname text, u_mobile text, u_is_admin integer u_created_at text)
            """)
        self.cursor.execute("delete from t_user where lower(u_username) = 'admin'")
        self.cursor.execute(
            """
            insert into t_user (u_id, u_username, u_password, u_nickname, u_mobile, u_is_admin, u_created_at)
            values ('0', 'admin', '5f4dcc3b5aa765d61d8327deb882cf99', '管理员', '', 1, datetime('now', 'localtime'))
            """)
        self.conn.commit()

    def __del__(self):
        if hasattr(self, 'cursor') and self.cursor:
            self.cursor.close()
        if hasattr(self, 'conn') and self.conn:
            self.conn.close()

    def to_user(self, row: sqlite3.Row) -> User:
        row_dict = dict(row)
        return User.restore_user(
            id=row_dict.get("u_id", ""),
            username=row_dict.get("u_username", ""),
            password=row_dict.get("u_password", ""),
            nickname=row_dict.get("u_nickname", ""),
            mobile=row_dict.get("u_mobile", ""),
            is_admin=bool(row_dict.get("u_is_admin", 0)),
            created_at=value_utility.to_datetime_req(row_dict.get("u_created_at", ""))
        )

    def insert(self, entity: User) -> None:
        self.cursor.execute("""
              insert into t_user (u_id, u_username, u_password, u_nickname, u_mobile, u_is_admin, u_created_at)
              values (?, ?, ?, ?, ?, ?, ?)
              """, (
            entity.id,
            entity.username,
            entity.password,
            entity.nickname,
            entity.mobile,
            int(entity.is_admin),
            entity.created_at
        ))
        self.conn.commit()

    def update(self, entity: User) -> None:
        self.cursor.execute("""
              update t_user set u_username = ?, u_password = ?, u_nickname = ?, u_mobile = ?, u_is_admin  = ?,
              u_created_at = ? where u_id = ?
              """, (
            entity.username,
            entity.password,
            entity.nickname,
            entity.mobile,
            int(entity.is_admin),
            entity.created_at,
            entity.id
        ))
        self.conn.commit()

    def delete_by_id(self, id: str) -> None:
        self.cursor.execute("delete from t_user where u_id = ?", (id,))
        # self.cursor.execute("delete from t_user where u_id = ?", [id])
        self.conn.commit()

    def select_by_id(self, id: str) -> User | None:
        row = self.cursor.execute(
            "select * from t_user where u_id = ?",(id,)).fetchone()
        if row:
            return self.to_user(row)
        return None

    def select_by_id_req(self, id: str) -> User:
        user = self.select_by_id(id)
        if not user:
            raise RepositoryException(f"用户 {id} 不存在")
        return user

    def select_by_ids(self, ids: list[str]) -> list[User]:
        if not ids or len(ids) == 0:
            return []
        placeholders = ",".join("?" for _ in ids)
        rows = self.cursor.execute(
            f"select * from t_user where u_id in ({placeholders})", ids
        ).fetchall()
        return [self.to_user(row) for row in rows]

    def select_by_username(self, username: str) -> User | None:
        row = self.cursor.execute(
            "select * from t_user where u_username = ?", (username,)
        ).fetchone()
        if row:
            return self.to_user(row)
        return None

    # ==================================================================================================================

    async def is_username_unique(self, username: str) -> bool:
        row = self.cursor.execute(
            "select 1 from t_user where u_username = ?", (username,)
        ).fetchone()
        return row is None

    async def is_nickname_unique(self, nickname: str) -> bool:
        row = self.cursor.execute(
            "select 1 from t_user where u_nickname = ?", (nickname,)
        ).fetchone()
        return row is None

    async def is_mobile_unique(self, mobile: str) -> bool:
        row = self.cursor.execute(
            "select 1 from t_user where u_mobile = ?", (mobile,)
        ).fetchone()
        return row is None

    # ==================================================================================================================

    def to_user_dto(self, row: sqlite3.Row) -> UserDto:
        row_dict = dict(row)
        return UserDto(
            id=row_dict.get("u_id", ""),
            username=row_dict.get("u_username", ""),
            # password=row_dict.get("u_password", ""),
            nickname=row_dict.get("u_nickname", ""),
            mobile=row_dict.get("u_mobile", ""),
            is_admin=bool(row_dict.get("u_is_admin", 0)),
            created_at=value_utility.to_datetime_req(row_dict.get("u_created_at", ""))
        )

    def query_by_id(self, id: str) -> UserDto | None:
        row = self.cursor.execute(
            "select * from t_user where u_id = ?", (id,)).fetchone()
        if row:
            return self.to_user_dto(row)
        return None

    def query_by_id_req(self, id: str) -> UserDto:
        user = self.query_by_id(id)
        if not user:
            raise QueryException(f"用户 {id} 不存在")
        return user

    def create_query_criteria_sql(self, criteria: UserQueryCriteria) -> tuple[str, list]:
        sqls = []
        params = []
        if criteria.keyword:
            sqls.append(" u_username like ? and u_nickname like ? ")
            params.append(f"%{criteria.keyword}%")
            params.append(f"%{criteria.keyword}%")
        if not sqls:
            return "", []
        else:
            return " where " + " and ".join(sqls), params

    def create_query_sort_sql(self, *sorts: UserQuerySort) -> str:
        sqls = []
        if sorts:
            for sort in sorts:
                match sort:
                    case UserQuerySort.CreateAtAsc:
                        sqls.append("u_created_at asc")
                    case UserQuerySort.CreateAtDesc:
                        sqls.append("u_created_at desc")
                    case UserQuerySort.UsernameAsc:
                        sqls.append("u_username desc")
                    case UserQuerySort.UsernameDesc:
                        sqls.append("u_username desc")

        if not sqls:
            return " order by u_created_at desc, u_id desc"
        else:
            sqls.append("u_id desc")
            return " order by " + ", ".join(sqls)

    def query_count(self, criteria: UserQueryCriteria) -> int:
        criteria_sql, params = self.create_query_criteria_sql(criteria)
        row = self.cursor.execute(f"select count(*) from t_user {criteria_sql}", params).fetchone()
        return row[0] if row else 0

    def query(self, criteria: UserQueryCriteria, *sorts: UserQuerySort) -> list[UserDto]:
        criteria_sql, params = self.create_query_criteria_sql(criteria)
        sort_sql = self.create_query_sort_sql(*sorts)
        rows = self.cursor.execute(
            f"select * from t_user {criteria_sql} {sort_sql}", params
        ).fetchall()
        return [self.to_user_dto(row) for row in rows]

    def query_by_page(self, page_num: int, page_size: int, criteria: UserQueryCriteria, *sorts: UserQuerySort) -> list[UserDto]:
        criteria_sql, params = self.create_query_criteria_sql(criteria)
        sort_sql = self.create_query_sort_sql(*sorts)
        offset = (page_num - 1) * page_size
        rows = self.cursor.execute(
            f"select * from t_user {criteria_sql} {sort_sql} limit ? offset ?",
            params + [page_size, offset]
        ).fetchall()
        return [self.to_user_dto(row) for row in rows]

File: dotnet/src/adapter/driven/UserPersistenceAdapter.cs
using System.Dynamic;

using Dapper;

using BackendExample.Application;
using BackendExample.Domain;
using BackendExample.Utility;

using Microsoft.Data.Sqlite;

namespace BackendExample.Adapter.Driven
{
    /// <summary>
    /// 用户持久化 Adapter
    /// </summary>
    public class UserPersistenceAdapter : UserRepository, UserUniqueSpecification, UserQueryHandler
    {
        private readonly string connectionString;

        public UserPersistenceAdapter(IConfiguration configuration)
        {
            this.connectionString = configuration.GetConnectionString("SQLite") ?? string.Empty;
            using var conn = new SqliteConnection(this.connectionString);
            conn.Open();
            conn.Execute("""
                create table if not exists t_user
                (
                    u_id text primary key,
                    u_username text,
                    u_password text,
                    u_nickname text,
                    u_mobile text,
                    u_is_admin integer,
                    u_created_at text
                )
                """);
            conn.Execute("delete from t_user where lower(u_username) = 'admin'");
            conn.Execute($"""
                insert into t_user
                (u_id, u_username, u_password, u_nickname, u_mobile, u_is_admin, u_created_at)
                values
                ('0', 'admin', '{CryptoUtility.EncodeMd5("password")}', '管理员', '', 1, datetime('now', 'localtime'))
                """);
        }

        private User ToUser(dynamic result)
        {
            return User.RestoreUser(
                result.u_id,
                result.u_username,
                result.u_password,
                result.u_nickname,
                result.u_mobile,
                ValueUtility.ToBoolOrDefault(result.u_is_admin.ToString(), false),
                ValueUtility.ToDateTimeOrDefault(result.u_created_at.ToString(), DateTime.MinValue)
                );
        }

        public async Task Insert(User entity)
        {
            using var conn = new SqliteConnection(this.connectionString);
            conn.Open();
            await conn.ExecuteAsync("""
                insert into t_user
                (u_id, u_username, u_password, u_nickname, u_mobile, u_is_admin, u_created_at)
                values
                (@id, @username, @password, @nickname, @mobile, @isAdmin, @createdAt)
                """, new
            {
                id = entity.Id,
                username = entity.Username,
                password = entity.Password,
                nickname = entity.Nickname,
                mobile = entity.Mobile,
                isAdmin = entity.IsAdmin ? 1 : 0,
                createdAt = entity.CreatedAt.ToString("yyyy-MM-dd HH:mm:ss")
            });
        }

        public async Task Update(User entity)
        {
            using var conn = new SqliteConnection(this.connectionString);
            conn.Open();
            await conn.ExecuteAsync("""
                update t_user
                set
                    u_username = @username,
                    u_password = @password,
                    u_nickname = @nickname,
                    u_mobile = @mobile,
                    u_is_admin = @isAdmin,
                    u_created_at = @createdAt
                where
                    u_id = @id
                """, new
            {
                id = entity.Id,
                username = entity.Username,
                password = entity.Password,
                nickname = entity.Nickname,
                mobile = entity.Mobile,
                isAdmin = entity.IsAdmin ? 1 : 0,
                createdAt = entity.CreatedAt.ToString("yyyy-MM-dd HH:mm:ss")
            });
        }


        public async Task DeleteById(string id)
        {
            if (ValueUtility.IsBlank(id))
                return;
            using var conn = new SqliteConnection(this.connectionString);
            conn.Open();
            await conn.ExecuteAsync("delete from t_user where u_id = @id", new { id });
        }

        public async Task<User?> SelectById(string id)
        {
            if (ValueUtility.IsBlank(id))
                return null;
            using var conn = new SqliteConnection(this.connectionString);
            conn.Open();
            dynamic? result = await conn.QuerySingleOrDefaultAsync("select * from t_user where u_id = @id", new { id });
            if (result == null)
                return null;
            return ToUser(result);
        }

        public async Task<User> SelectByIdReq(string id)
        {
            User? entity = await SelectById(id);
            if (entity == null)
                throw new PersistenceException($"用户 {id} 不存在");
            return entity;
        }

        public async Task<IList<User>> SelectByIds(List<string> ids)
        {
            IList<User> list = new List<User>();
            if (ids == null || ids.Count == 0)
                return list;

            using var conn = new SqliteConnection(this.connectionString);
            conn.Open();
            IEnumerable<dynamic> result = await conn.QueryAsync("select * from t_user where u_id in @ids", new { ids });
            foreach (dynamic item in result)
            {
                User entity = ToUser(item);
                list.Add(entity);
            }
            return list;
        }

        public async Task<User?> SelectByUsername(string username)
        {
            if (ValueUtility.IsBlank(username))
                return null;
            using var conn = new SqliteConnection(this.connectionString);
            conn.Open();
            dynamic? result = await conn.QuerySingleOrDefaultAsync("select * from t_user where u_username = @username",
                new { username });
            if (result == null)
                return null;
            return ToUser(result);
        }

        //==============================================================================================================
        // UserUniqueSpecification

        public async Task<bool> IsUsernameUnique(string username)
        {
            if (ValueUtility.IsBlank(username))
                throw new PersistenceException("参数 username 不能为空");
            using var conn = new SqliteConnection(this.connectionString);
            conn.Open();
            return await conn.ExecuteScalarAsync<int>(
                "select count(1) from t_user where lower(u_username) = lower(@username)",
                new { username }
                ) == 0;
        }

        public async Task<bool> IsNicknameUnique(string nickname)
        {
            if (ValueUtility.IsBlank(nickname))
                throw new PersistenceException("参数 nickname 不能为空");
            using var conn = new SqliteConnection(this.connectionString);
            conn.Open();
            return await conn.ExecuteScalarAsync<int>(
                "select count(1) from t_user where lower(u_nickname) = lower(@nickname)",
                new { nickname }
                ) == 0;
        }

        public async Task<bool> IsMobileUnique(string mobile)
        {
            if (ValueUtility.IsBlank(mobile))
                throw new PersistenceException("参数 mobile 不能为空");
            using var conn = new SqliteConnection(this.connectionString);
            conn.Open();
            return await conn.ExecuteScalarAsync<int>(
                "select count(1) from t_user where lower(u_mobile) = lower(@mobile)",
                new { mobile }
                ) == 0;
        }

        //==============================================================================================================
        // UserQueryHandler

        private UserDto ToUserDto(dynamic result)
        {
            return new UserDto(
                result.u_id,
                result.u_username,
                //result.u_password,
                result.u_nickname,
                result.u_mobile,
                ValueUtility.ToBoolOrDefault(result.u_is_admin.ToString(), false),
                ValueUtility.ToDateTimeOrDefault(result.u_created_at.ToString(), DateTime.MinValue)
                );
        }

        private string BuildQueryCriteria(UserQueryCriteria? criteria, out dynamic @params)
        {
            @params = new ExpandoObject();

            if (criteria == null)
                return string.Empty;

            IList<string> sqls = new List<string>();
            if (!ValueUtility.IsBlank(criteria.Keyword))
            {
                sqls.Add("u_username like @Keyword or u_nickname like @keyword");
                @params.keyword = criteria.Keyword;
            }

            if (sqls.Count > 0)
                return " where " + string.Join(" and ", sqls);

            return string.Empty;
        }

        private string BuildQuerySort(params UserQuerySort[] sorts)
        {
            IList<string> sqls = new List<string>();
            foreach (UserQuerySort sort in sorts)
            {
                switch (sort)
                {
                    case UserQuerySort.CREATED_AT_ASC:
                        sqls.Add("u_created_at asc");
                        break;
                    case UserQuerySort.CREATED_AT_DESC:
                        sqls.Add("u_created_at desc");
                        break;
                    case UserQuerySort.USERNAME_ASC:
                        sqls.Add("u_username asc");
                        break;
                    case UserQuerySort.USERNAME_DESC:
                        sqls.Add("u_username desc");
                        break;
                }
            }

            if (sqls.Count == 0)
                sqls.Add("u_created_at desc");

            return " order by " + string.Join(", ", sqls);
        }

        public async Task<UserDto?> QueryById(string id)
        {
            if (ValueUtility.IsBlank(id))
                return null;
            using var conn = new SqliteConnection(this.connectionString);
            conn.Open();
            dynamic? result = await conn.QuerySingleOrDefaultAsync("select * from t_user where u_id = @id", new { id });
            if (result == null)
                return null;
            return ToUserDto(result);
        }

        public async Task<UserDto> QueryByIdReq(string id)
        {
            UserDto? dto = await QueryById(id);
            if (dto == null)
                throw new PersistenceException($"用户 {id} 不存在");
            return dto;
        }

        public async Task<int> QueryCount(UserQueryCriteria? criteria)
        {
            dynamic @params;
            string crieriaSql = BuildQueryCriteria(criteria, out @params);
            using var conn = new SqliteConnection(this.connectionString);
            conn.Open();
            return await conn.ExecuteScalarAsync<int>("select count(*) from t_user" + crieriaSql, (object)@params);
        }

        public async Task<IList<UserDto>> Query(UserQueryCriteria? criteria, params UserQuerySort[] sorts)
        {
            dynamic @params;
            string crieriaSql = BuildQueryCriteria(criteria, out @params);
            string sortSql = BuildQuerySort(sorts);
            using var conn = new SqliteConnection(this.connectionString);
            conn.Open();
            IEnumerable<dynamic> result = await conn.QueryAsync("select * from t_user" + crieriaSql + sortSql, (object)@params);
            IList<UserDto> list = new List<UserDto>();
            foreach (dynamic item in result)
            {
                UserDto dto = ToUserDto(item);
                list.Add(dto);
            }
            return list;
        }

        public async Task<IList<UserDto>> QueryByPage(int pageNum, int pageSize, UserQueryCriteria? criteria,
            params UserQuerySort[] sorts)
        {
            dynamic @params;
            string crieriaSql = BuildQueryCriteria(criteria, out @params);
            string sortSql = BuildQuerySort(sorts);
            @params.limitCount = pageSize;
            @params.limitOffset = (pageNum - 1) * pageSize;

            using var conn = new SqliteConnection(this.connectionString);
            conn.Open();
            IEnumerable<dynamic> result = await conn.QueryAsync(
                "select * from t_user" + crieriaSql + sortSql + "limit @limitCount offset @limitOffset",
                (object)@params
                );
            IList<UserDto> list = new List<UserDto>();
            foreach (dynamic item in result)
            {
                UserDto dto = ToUserDto(item);
                list.Add(dto);
            }
            return list;
        }
    }
}
